### ISSUE TITLE
For 'sound null safety support', teach pub how to distinguish code generators from other libraries?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/pub-dev#4680]
pub.dev lists [reflectable](https://pub.dev/packages/reflectable/score) 3.0.0 as _not_ supporting sound null safety. It seems likely that this is based on the fact that at least one of the libraries in this package contains `@dart = 2.9`, or it occurs because at least one library imports a library that isn't considered null safe.

However, this package consists of some libraries that implement a code generator, and other libraries that are imported by client packages using reflectable. The libraries that are imported into client programs are fully migrated to null safety, it is only the code generator that depends on some packages/libraries that aren't yet fully migrated (hence, a couple of libraries in the code generator contain `@dart = 2.9`).

There are 69 tests running with sound null safety, so it's well established that client programs can be fully null safe.

Is there a way to make this distinction, indicating that a package with both client-imported libraries and "tool" libraries can work as a fully null safe package, even though it contains one or more tools that haven't yet finished the migration?

**Comment by sigurdm** [comment]
Yeah - we really should distinguish "tools/code-generators" somehow.

We need to find out the right signal that a package is a code-generator.

* No public libraries?
* a specific entry in the pubspec?
* ...?

**Comment by sigurdm** [comment]
> It seems likely that this is based on the fact that at least one of the libraries in this package contains @dart = 2.9, or it occurs because at least one library imports a library that isn't considered null safe.

Just fictually pana/pub.dev actually gives a reason at https://pub.dev/packages/reflectable/score

```
Package is not null safe
Because:

reflectable that depends on:
build_runner_core that doesn't opt in to null safety
```

(It just gives the first reason it finds, and `@dart = 2.9` is another reason it would not mark the package null-safety).

**Comment by eernstg** [comment]
Thanks! Btw, I expect that there will soon be a null safe `build`, and then reflectable can opt out of this issue, independently of any solutions we might end up having. ;-)

**Comment by buzzware** [comment]
It seems this issue is holding up a chain of packages and apps including mine from becoming null safe.

ie my app ->  dart_json_mapper -> reflectable -> build_runner_core

I believe they have null safety in their code, but don't get the necessary badge due to this issue.

**Comment by eernstg** [comment]
`build` is now available in a version with null safety, and the reflectable code generator is currently being migrated accordingly.

**Comment by cedvdb** [comment]
dev dependencies shouldn't have an impact on whether or not a package is or isn't null safe. Should they have any impact at all anyway in the score ?

**Comment by buzzware** [comment]
I'm only here because of this comment https://github.com/k-paxian/dart-json-mapper/issues/97#issuecomment-826602381
Sounds like the work is underway anyhow.

**Comment by eernstg** [comment]
`dev_dependencies` are described as follows [here]():

> Dependencies fall into one of two types. Regular dependencies are listed
> under `dependencies:`—these are packages that anyone using your package
> will also need. Dependencies that are only needed in the development phase
> of the package itself are listed under dev_dependencies.

It isn't actually true that a code generator is only used in the development of the package itself, so presumably the code generators weren't taken into account when this was written.

However, when [this](https://github.com/dart-lang/sdk/issues/45839) is fixed we should be able to land [this](https://github.com/google/reflectable.dart/pull/248), which would make the worries about dependencies more theoretical.

**Comment by sigurdm** [comment]
Perhaps being a "tool" could be part of the top-level platforms:

platforms:
  tool

indicating this package is meant to be run in the dev-environment, not as a runtime dependency.

Or we could have another top-level marker for this.

Alternatively we could try to deduct this by some heuristic:

Either a tool has:

* Setup for the build_runner tool chain
* or has `executables` in the pubspec
